sonarts-core/src/se/stateTransitions.ts

```diff
@@ -0,0 +1,114 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as ts from "typescript";
+import * as tsutils from "tsutils";
+import { createLiteralSymbolicValue, createUnknownSymbolicValue, createUndefinedSymbolicValue } from "./symbolicValues";
+import { ProgramState } from "./programStates";
+
+export function applyExecutors(programPoint: ts.Node, state: ProgramState, program: ts.Program): ProgramState {
+  const { parent } = programPoint;
+
+  // TODO is there a better way to handle this?
+  // special case: `let x;`
+  if (parent && tsutils.isVariableDeclaration(parent) && parent.name === programPoint) {
+    return variableDeclaration(parent, state, program);
+  }
+
+  if (tsutils.isNumericLiteral(programPoint)) {
+    return numeralLiteral(programPoint, state, program);
+  }
+
+  if (tsutils.isIdentifier(programPoint)) {
+    return identifier(programPoint, state, program);
+  }
+
+  if (tsutils.isBinaryExpression(programPoint)) {
+    return binaryExpression(programPoint, state, program);
+  }
+
+  if (tsutils.isVariableDeclaration(programPoint)) {
+    return variableDeclaration(programPoint, state, program);
+  }
+
+  if (tsutils.isCallExpression(programPoint)) {
+    return callExpression(programPoint, state);
+  }
+
+  return state;
+}
+
+function identifier(identifier: ts.Identifier, state: ProgramState, program: ts.Program) {
+  const symbol = program.getTypeChecker().getSymbolAtLocation(identifier);
+  if (symbol) {
+    const sv = state.sv(symbol) || createUnknownSymbolicValue();
+    return state.pushSV(sv);
+  }
+  return state;
+}
+
+function numeralLiteral(literal: ts.NumericLiteral, state: ProgramState, _program: ts.Program) {
+  return state.pushSV(createLiteralSymbolicValue(literal.text));
+}
+
+function binaryExpression(expression: ts.BinaryExpression, state: ProgramState, program: ts.Program) {
+  if (expression.operatorToken.kind === ts.SyntaxKind.EqualsToken) {
+    let [value, nextState] = state.popSV();
+    const { getSymbolAtLocation } = program.getTypeChecker();
+    const variable = getSymbolAtLocation(expression.left as ts.Identifier);
+    if (!variable) {
+      return nextState;
+    }
+
+    if (!value) {
+      throw "Assignment without value";
+    }
+    nextState = nextState.pushSV(value);
+  
+    return nextState.setSV(variable, value);
+  }
+  return state;
+}
+
+function variableDeclaration(declaration: ts.VariableDeclaration, state: ProgramState, program: ts.Program) {
+  if (tsutils.isIdentifier(declaration.name)) {
+    let [value, nextState] = state.popSV();
+    const { getSymbolAtLocation } = program.getTypeChecker();
+    const variable = getSymbolAtLocation(declaration.name);    
+    if (!variable) {
+      return nextState;
+    }
+
+    if (!value) {
+      value = createUndefinedSymbolicValue();
+    }
+    if (!nextState.hasEmptyStack()) {
+      throw "Non empty stack at the end of declaration " + declaration.getText();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Throwing plain strings (not instances of Error) gives no stack traces [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS3696)